### DISCUSSION TITLE
Alert if Agg Count is 1

### DISCUSSION THREAD

**Discussion by chivasblue** [jertel/elastalert2#791] [Q&A] [answered]
Hello,
I am new to elastalert2 and was wondering if it can be used for my particular case. I have log lines like below which are written to elasticsearch.

_timestamp,12345,REQ,myfunctioncall
timestamp,12345,RES,myfunctioncall_

so each function call request (REQ) has a corresponding response (RES) with same id (12345), Can I create a alert in case there is no RES (response call) to a id within a certain timeframe, is it possible and can someone point me to the example?

Thanks in advance!

**Comment by jertel** [comment, marked as answer]
The [cardinality](https://elastalert2.readthedocs.io/en/latest/ruletypes.html#cardinality) rule might work for you. 
```
cardinality_field: function.name
query_key: id
```

**Reply by chivasblue** [reply]
@jertel thanks for your reply, let me take a look into it.

**Reply by chivasblue** [reply]
@jertel I tried the cardinality rule but it does not seem to match, I am looking for something on lines of below search query which works fine.

```
GET logs-test*/_search
{
  "aggs": {
    "id_count": {
      "terms": {
        "field": "id"
      },
         "aggs": {
            "count_filter": {
               "bucket_selector": {
                  "buckets_path": {
                     "the_doc_count": "_count"
                  },
                  "script": "params.the_doc_count == 1"
               }
            }
         }
    }
  },
  "size": 0,
  "query": {
    "bool": {
      "filter": [
        {
          "match_phrase": {
            "body.function": "myfunctioncall"
          }
        },
        {
          "range": {
            "timestamp": {
              "gte": "2022-03-30T17:00:00.000Z",
              "lte": "2022-03-31T16:59:59.999Z"
            }
          }
        }
      ]
    }
  }
}
```
Any pointers?

**Reply by jertel** [reply]
Please be more specific when you say the cardinality rule "didn't match".

The query you suggested is simpler than what you originally requested, which means you could use the `cardinality` rule type, `metric_aggregation` rule type, and probably the `flatline` rule type (with  query_key). There are several examples of rules in the `examples` folder.

**Reply by chivasblue** [reply]
@jertel Sorry for not providing more details, below is my yaml file for cardinality rule

```
index: logs-test*

name: No Response

type: cardinality

cardinality_field: function

min_cardinality: 1
query_key: id

timeframe:
  minutes: 15

filter:
- term:
    function: "myfunctioncall"

alert:
- "debug"
```

My log lines:

```
{"timestamp":"2022-03-31T14:20:46.860Z","id":1,"function":"myfunctioncall","type":"REQ"}
{"timestamp":"2022-03-31T14:20:46.860Z","id":2,"function":"myfunctioncall","type":"REQ"}
{"timestamp":"2022-03-31T14:20:00.860Z","id":3,"function":"myfunctioncall","type":"REQ"}
{"timestamp":"2022-03-31T14:10:30.860Z","id":4,"function":"myfunctioncall","type":"REQ"}
```
Output when I test the rule:

_elastalert-test-rule rules/no_response.yaml 
INFO:elastalert:Note: In debug mode, alerts will be logged to console but NOT actually sent.
            To send them but remain verbose, use --verbose instead.
Got 4 hits from the last 0 day

Available terms in first hit:
	message
	timestamp
	function
	host.name
	id
	fields.log_type
	type
	@timestamp

INFO:elastalert:Note: In debug mode, alerts will be logged to console but NOT actually sent.
                To send them but remain verbose, use --verbose instead.
1 rules loaded
INFO:apscheduler.scheduler:Adding job tentatively -- it will be properly scheduled when the scheduler starts
INFO:elastalert:Queried rule No Response from 2022-03-31 21:05 +07 to 2022-03-31 21:06 +07: 0 / 0 hits
INFO:elastalert:Queried rule No Response from 2022-03-31 21:06 +07 to 2022-03-31 21:07 +07: 0 / 0 hits
INFO:elastalert:Queried rule No Response from 2022-03-31 21:07 +07 to 2022-03-31 21:08 +07: 0 / 0 hits
INFO:elastalert:Queried rule No Response from 2022-03-31 21:08 +07 to 2022-03-31 21:09 +07: 0 / 0 hits
INFO:elastalert:Queried rule No Response from 2022-03-31 21:09 +07 to 2022-03-31 21:10 +07: 0 / 0 hits
INFO:elastalert:Queried rule No Response from 2022-03-31 21:10 +07 to 2022-03-31 21:11 +07: 0 / 0 hits
INFO:elastalert:Queried rule No Response from 2022-03-31 21:11 +07 to 2022-03-31 21:12 +07: 0 / 0 hits
INFO:elastalert:Queried rule No Response from 2022-03-31 21:12 +07 to 2022-03-31 21:13 +07: 0 / 0 hits
INFO:elastalert:Queried rule No Response from 2022-03-31 21:13 +07 to 2022-03-31 21:14 +07: 0 / 0 hits
INFO:elastalert:Queried rule No Response from 2022-03-31 21:14 +07 to 2022-03-31 21:15 +07: 0 / 0 hits
INFO:elastalert:Queried rule No Response from 2022-03-31 21:15 +07 to 2022-03-31 21:16 +07: 0 / 0 hits
INFO:elastalert:Queried rule No Response from 2022-03-31 21:16 +07 to 2022-03-31 21:17 +07: 0 / 0 hits
INFO:elastalert:Queried rule No Response from 2022-03-31 21:17 +07 to 2022-03-31 21:18 +07: 0 / 0 hits
INFO:elastalert:Queried rule No Response from 2022-03-31 21:18 +07 to 2022-03-31 21:19 +07: 0 / 0 hits
INFO:elastalert:Queried rule No Response from 2022-03-31 21:19 +07 to 2022-03-31 21:20 +07: 4 / 4 hits
INFO:elastalert:Queried rule No Response from 2022-03-31 21:20 +07 to 2022-03-31 21:20 +07: 0 / 0 hits

Would have written the following documents to writeback index (default is elastalert_status):

elastalert_status - {'rule_name': 'No Response', 'endtime': datetime.datetime(2022, 3, 31, 14, 20, 42, 817773, tzinfo=tzutc()), 'starttime': datetime.datetime(2022, 3, 31, 14, 5, 33, 817773, tzinfo=tzutc()), 'matches': 0, 'hits': 4, '@timestamp': datetime.datetime(2022, 3, 31, 14, 20, 43, 202788, tzinfo=tzutc()), 'time_taken': 0.20719194412231445}_

I think the matches field should have some value other than 0 to send the alert, correct?

Thanks for your help.

**Reply by jertel** [reply]
This data you've just provided is very helpful! It should be a model for people requesting support.

The `cardinality_field` should be `type`, not `function`. I used `function.name` in my first response since I didn't know what the field names were in your original post. But my intention was the REQ and RES were function.names. Since that field name is actually called `type` use that instead. Also, change the `min_cardinality` to `2`, since you want the alert to trigger when that minimum threshold isn't satisfied. Setting it to 1 is telling ElastAlert that all is well as long as you have at least one unique type value for that ID, and that's not what you want. You want to alert (all is NOT well) if there's only one cardinality.

**Reply by chivasblue** [reply]
@jertel no luck yet after making the change :(

cat rules/no_response.yaml
```

index: logs-test*

name: No Response

type: cardinality

cardinality_field: type

min_cardinality: 2
query_key: id

timeframe:
  minutes: 15

filter:
- term:
    function: "myfunctioncall"

alert:
- "debug"
```
Log Lines (have added last line with function: yfunctioncall purposefully)
```
{"timestamp":"2022-03-31T14:53:46.860Z","id":1,"function":"myfunctioncall","type":"REQ"}
{"timestamp":"2022-03-31T14:53:46.860Z","id":2,"function":"myfunctioncall","type":"REQ"}
{"timestamp":"2022-03-31T14:53:00.860Z","id":3,"function":"myfunctioncall","type":"REQ"}
{"timestamp":"2022-03-31T14:53:30.860Z","id":3,"function":"myfunctioncall","type":"RES"}
{"timestamp":"2022-03-31T14:53:30.860Z","id":4,"function":"yfunctioncall","type":"REQ"}
```

 elastalert-test-rule rules/no_response.yaml
INFO:elastalert:Note: In debug mode, alerts will be logged to console but NOT actually sent.
            To send them but remain verbose, use --verbose instead.
Got 4 hits from the last 0 day

Available terms in first hit:
	message
	timestamp
	function
	host.name
	id
	fields.log_type
	type
	@timestamp

INFO:elastalert:Note: In debug mode, alerts will be logged to console but NOT actually sent.
                To send them but remain verbose, use --verbose instead.
1 rules loaded
INFO:apscheduler.scheduler:Adding job tentatively -- it will be properly scheduled when the scheduler starts
INFO:elastalert:Queried rule No Response from 2022-03-31 22:11 +07 to 2022-03-31 22:12 +07: 0 / 0 hits
INFO:elastalert:Queried rule No Response from 2022-03-31 22:12 +07 to 2022-03-31 22:13 +07: 0 / 0 hits
INFO:elastalert:Queried rule No Response from 2022-03-31 22:13 +07 to 2022-03-31 22:14 +07: 0 / 0 hits
INFO:elastalert:Queried rule No Response from 2022-03-31 22:14 +07 to 2022-03-31 22:15 +07: 0 / 0 hits
INFO:elastalert:Queried rule No Response from 2022-03-31 22:15 +07 to 2022-03-31 22:16 +07: 0 / 0 hits
INFO:elastalert:Queried rule No Response from 2022-03-31 22:16 +07 to 2022-03-31 22:17 +07: 0 / 0 hits
INFO:elastalert:Queried rule No Response from 2022-03-31 22:17 +07 to 2022-03-31 22:18 +07: 0 / 0 hits
INFO:elastalert:Queried rule No Response from 2022-03-31 22:18 +07 to 2022-03-31 22:19 +07: 0 / 0 hits
INFO:elastalert:Queried rule No Response from 2022-03-31 22:19 +07 to 2022-03-31 22:20 +07: 0 / 0 hits
INFO:elastalert:Queried rule No Response from 2022-03-31 22:20 +07 to 2022-03-31 22:21 +07: 0 / 0 hits
INFO:elastalert:Queried rule No Response from 2022-03-31 22:21 +07 to 2022-03-31 22:22 +07: 0 / 0 hits
INFO:elastalert:Queried rule No Response from 2022-03-31 22:22 +07 to 2022-03-31 22:23 +07: 0 / 0 hits
INFO:elastalert:Queried rule No Response from 2022-03-31 22:23 +07 to 2022-03-31 22:24 +07: 4 / 4 hits
INFO:elastalert:Queried rule No Response from 2022-03-31 22:24 +07 to 2022-03-31 22:25 +07: 0 / 0 hits
INFO:elastalert:Queried rule No Response from 2022-03-31 22:25 +07 to 2022-03-31 22:26 +07: 0 / 0 hits
INFO:elastalert:Queried rule No Response from 2022-03-31 22:26 +07 to 2022-03-31 22:26 +07: 0 / 0 hits

Would have written the following documents to writeback index (default is elastalert_status):

elastalert_status - {'rule_name': 'No Response', 'endtime': datetime.datetime(2022, 3, 31, 15, 26, 59, 803212, tzinfo=tzutc()), 'starttime': datetime.datetime(2022, 3, 31, 15, 11, 50, 803212, tzinfo=tzutc()), 'matches': 0, 'hits': 4, '@timestamp': datetime.datetime(2022, 3, 31, 15, 27, 0, 157577, tzinfo=tzutc()), 'time_taken': 0.1840670108795166}


Appreciate all your help!

**Reply by jertel** [reply]
Did you wait for the full timeframe to elapse, during your test?

**Reply by chivasblue** [reply]
@jertel I do not quite understand what you mean by waiting for full timeframe to elapse, anyways I made the below changes to cardinality yaml file (timeframe from 15 mins to 1 min)

```
index: logs-test*

name: No Response

type: cardinality

cardinality_field: type

min_cardinality: 2
query_key: id

timeframe:
  minutes: 1

filter:
- term:
    function: "myfunctioncall"

alert:
- "debug"

```

My log file (btw my timezone is ICT)

```
{"timestamp":"2022-04-01T04:46:46.860Z","id":1,"function":"myfunctioncall","type":"REQ"}
{"timestamp":"2022-04-01T04:46:46.860Z","id":2,"function":"myfunctioncall","type":"REQ"}
{"timestamp":"2022-04-01T04:46:00.860Z","id":3,"function":"myfunctioncall","type":"REQ"}
{"timestamp":"2022-04-01T04:46:30.860Z","id":3,"function":"myfunctioncall","type":"RES"}
{"timestamp":"2022-04-01T04:46:30.860Z","id":4,"function":"yfunctioncall","type":"REQ"}
```

elastalert-test-rule rules/no_response.yaml
INFO:elastalert:Note: In debug mode, alerts will be logged to console but NOT actually sent.
            To send them but remain verbose, use --verbose instead.
Got 4 hits from the last 0 day

Available terms in first hit:
	message
	timestamp
	function
	host.name
	id
	fields.log_type
	type
	@timestamp

INFO:elastalert:Note: In debug mode, alerts will be logged to console but NOT actually sent.
                To send them but remain verbose, use --verbose instead.
1 rules loaded
INFO:apscheduler.scheduler:Adding job tentatively -- it will be properly scheduled when the scheduler starts
INFO:elastalert:Queried rule No Response from 2022-04-01 11:45 +07 to 2022-04-01 11:46 +07: **4 / 4 hits**
INFO:elastalert:Queried rule No Response from 2022-04-01 11:46 +07 to 2022-04-01 11:46 +07: 0 / 0 hits

Would have written the following documents to writeback index (default is elastalert_status):

elastalert_status - {'rule_name': 'No Response', 'endtime': datetime.datetime(2022, 4, 1, 4, 46, 45, 850998, tzinfo=tzutc()), 'starttime': datetime.datetime(2022, 4, 1, 4, 45, 45, 250998, tzinfo=tzutc()), 'matches': 0, 'hits': 4, '@timestamp': datetime.datetime(2022, 4, 1, 4, 46, 46, 48910, tzinfo=tzutc()), 'time_taken': 0.02754998207092285}

**So got hits the first time, but no matches. Ran again.**

elastalert-test-rule rules/no_response.yaml
INFO:elastalert:Note: In debug mode, alerts will be logged to console but NOT actually sent.
            To send them but remain verbose, use --verbose instead.
Got 4 hits from the last 0 day

Available terms in first hit:
	message
	timestamp
	function
	host.name
	id
	fields.log_type
	type
	@timestamp

INFO:elastalert:Note: In debug mode, alerts will be logged to console but NOT actually sent.
                To send them but remain verbose, use --verbose instead.
1 rules loaded
INFO:apscheduler.scheduler:Adding job tentatively -- it will be properly scheduled when the scheduler starts
INFO:elastalert:Queried rule No Response from 2022-04-01 11:46 +07 to 2022-04-01 11:47 +07: **0 / 0 hits**
INFO:elastalert:Queried rule No Response from 2022-04-01 11:47 +07 to 2022-04-01 11:47 +07: **0 / 0 hits**

Would have written the following documents to writeback index (default is elastalert_status):

elastalert_status - {'rule_name': 'No Response', 'endtime': datetime.datetime(2022, 4, 1, 4, 47, 3, 496340, tzinfo=tzutc()), 'starttime': datetime.datetime(2022, 4, 1, 4, 46, 2, 896340, tzinfo=tzutc()), 'matches': 0, 'hits': 0, '@timestamp': datetime.datetime(2022, 4, 1, 4, 47, 3, 695309, tzinfo=tzutc()), 'time_taken': 0.027435302734375}

No hit/matches.

So does this qualify as waiting for full timeframe to elapse?

**Reply by jertel** [reply]
> So does this qualify as waiting for full timeframe to elapse?

No. You will need to specify a `--start` param to the `elastalert-test-rule` command. Ex:

```
elastalert-test-rule --start 2022-03-31T14:25:00.000Z rules/no_response.yaml
```

This is required because in order for a `min_cardinality` check to work it must see the entire timeframe elapse starting from the first event is found. In your tests it is finding the first event successfully within 60 seconds of the time you execute the test command, but the test stops before the full 1 minute elapses. Adding the `--start` param to the test command line will tell the test to look further back in history to find a matching event and start the clock at that event.

**Reply by chivasblue** [reply]
@jertel I am still not able to get it right, my config file is below:


```
index: logs-test*

name: No Response

type: cardinality

cardinality_field: type

min_cardinality: 2
query_key: id

timeframe:
  minutes: 1

filter:
- term:
    function: "myfunctioncall"

alert:
- debug
```


My log file:
```
{"timestamp":"2022-04-06T10:41:46.860Z","id":1,"function":"myfunctioncall","type":"REQ"}
{"timestamp":"2022-04-06T10:41:46.860Z","id":1,"function":"myfunctioncall","type":"RES"}
{"timestamp":"2022-04-06T10:41:00.860Z","id":3,"function":"myfunctioncall","type":"REQ"}
{"timestamp":"2022-04-06T10:41:30.860Z","id":3,"function":"myfunctioncall","type":"RES"}
{"timestamp":"2022-04-06T10:41:30.860Z","id":4,"function":"yfunctioncall","type":"REQ"}
{"timestamp":"2022-04-06T10:41:30.860Z","id":4,"function":"yfunctioncall","type":"RES"}
```
Based on above log file I am not expecting any matches as every id has a REQ and corresponding RES. However when I test the rule:

elastalert-test-rule --start 2022-04-06T10:40:00.000Z no_response.yaml
INFO:elastalert:Note: In debug mode, alerts will be logged to console but NOT actually sent.
            To send them but remain verbose, use --verbose instead.
Got 4 hits from the last 0 day

Available terms in first hit:
	reqTime
	message
	respTime
	timestamp
	function
	host.name
	id
	fields.log_type
	type
	resTime
	@timestamp

INFO:elastalert:Note: In debug mode, alerts will be logged to console but NOT actually sent.
                To send them but remain verbose, use --verbose instead.
1 rules loaded
INFO:apscheduler.scheduler:Adding job tentatively -- it will be properly scheduled when the scheduler starts
INFO:elastalert:Queried rule No Response from 2022-04-06 17:40 +07 to 2022-04-06 17:41 +07: 0 / 0 hits
INFO:elastalert:Queried rule No Response from 2022-04-06 17:41 +07 to 2022-04-06 17:42 +07: 4 / 4 hits
INFO:elastalert:Queried rule No Response from 2022-04-06 17:42 +07 to 2022-04-06 17:42 +07: 0 / 0 hits
INFO:elastalert:Alert for No Response, 3 at 2022-04-06T10:42:11.929130Z:
INFO:elastalert:No Response

Less than 2 unique type(s) occurred since last alert or between 2022-04-06 17:41 +07 and 2022-04-06 17:42 +07

@timestamp: 2022-04-06T10:42:11.929130Z
id: 3
num_hits: 4
num_matches: 2

INFO:elastalert:Alert for No Response, 1 at 2022-04-06T10:42:11.929130Z:
INFO:elastalert:No Response

Less than 2 unique type(s) occurred since last alert or between 2022-04-06 17:41 +07 and 2022-04-06 17:42 +07

@timestamp: 2022-04-06T10:42:11.929130Z
id: 1
num_hits: 4
num_matches: 2


Would have written the following documents to writeback index (default is elastalert_status):

silence - {'exponent': 0, 'rule_name': 'No Response.3', '@timestamp': datetime.datetime(2022, 4, 6, 10, 42, 12, 139897, tzinfo=tzutc()), 'until': datetime.datetime(2022, 4, 6, 10, 43, 12, 139883, tzinfo=tzutc())}

silence - {'exponent': 0, 'rule_name': 'No Response.1', '@timestamp': datetime.datetime(2022, 4, 6, 10, 42, 12, 141104, tzinfo=tzutc()), 'until': datetime.datetime(2022, 4, 6, 10, 43, 12, 141094, tzinfo=tzutc())}

elastalert_status - {'rule_name': 'No Response', 'endtime': datetime.datetime(2022, 4, 6, 10, 42, 11, 929130, tzinfo=tzutc()), 'starttime': datetime.datetime(2022, 4, 6, 10, 40, tzinfo=tzutc()), 'matches': 2, 'hits': 4, '@timestamp': datetime.datetime(2022, 4, 6, 10, 42, 12, 141889, tzinfo=tzutc()), 'time_taken': 0.03763294219970703}

Why are there matches?

Thanks.

**Reply by jertel** [reply]
Have you tried running the rule in ElastAlert 2, instead of using the test command? If not I recommend doing so to help determine if there's still a timing issue that the test tool isn't able to overcome.

Also, are your field mappings adequate for these aggregations? Both `id` and `type` need to be aggregatable.

Last, I suggest enabling debug logging so you can see the raw request and responses being transmitted between EA and ES. The docs talk about the config options needing to increase logging output. You can also search this discussion forum.
